File: cpp/obsidianmesh/src/statistics.cpp
#include "obsidianmesh/core.hpp"
#include <algorithm>
#include <cmath>

namespace obsidianmesh {

// ---------------------------------------------------------------------------
// Core percentile function
// ---------------------------------------------------------------------------

int percentile(std::vector<int> values, int pct) {
  if (values.empty()) return 0;
  std::sort(values.begin(), values.end());
  int rank = (pct * static_cast<int>(values.size())) / 100;
  if (rank < 0) rank = 0;
  if (rank >= static_cast<int>(values.size())) rank = static_cast<int>(values.size()) - 1;
  return values[static_cast<size_t>(rank)];
}

// ---------------------------------------------------------------------------
// Descriptive statistics
// ---------------------------------------------------------------------------

double mean(const std::vector<double>& values) {
  if (values.empty()) return 0;
  double sum = 0;
  for (auto v : values) sum += v;
  return sum / static_cast<double>(values.size());
}

double variance(const std::vector<double>& values) {
  if (values.size() < 2) return 0;
  double avg = mean(values);
  double sum_sq = 0;
  for (auto v : values) {
    double diff = v - avg;
    sum_sq += diff * diff;
  }
  return sum_sq / static_cast<double>(values.size() - 1);
}

double stddev(const std::vector<double>& values) {
  return std::sqrt(variance(values));
}

double median(std::vector<double> values) {
  if (values.empty()) return 0;
  std::sort(values.begin(), values.end());
  size_t mid = values.size() / 2;
  if (values.size() % 2 == 0) {
    return (values[mid - 1] + values[mid]) / 2.0;
  }
  return values[mid];
}

// ---------------------------------------------------------------------------
// Response time tracker
// ---------------------------------------------------------------------------

ResponseTimeTracker::ResponseTimeTracker(int window_size)
    : window_size_(window_size > 0 ? window_size : 1000) {}

void ResponseTimeTracker::record(double duration_ms) {
  std::lock_guard lock(mu_);
  samples_.push_back(duration_ms);
  if (static_cast<int>(samples_.size()) > window_size_) {
    samples_.erase(samples_.begin());
  }
}

double ResponseTimeTracker::percentile_float(int pct) {
  if (samples_.empty()) return 0;
  auto cloned = samples_;
  std::sort(cloned.begin(), cloned.end());
  int rank = ((pct * static_cast<int>(cloned.size())) + 99) / 100;
  if (rank <= 0) rank = 1;
  if (rank > static_cast<int>(cloned.size())) rank = static_cast<int>(cloned.size());
  return cloned[static_cast<size_t>(rank - 1)];
}

double ResponseTimeTracker::p50() {
  std::lock_guard lock(mu_);
  return percentile_float(50);
}

double ResponseTimeTracker::p95() {
  std::lock_guard lock(mu_);
  return percentile_float(95);
}

double ResponseTimeTracker::p99() {
  std::lock_guard lock(mu_);
  return percentile_float(99);
}

double ResponseTimeTracker::average() {
  std::lock_guard lock(mu_);
  return mean(samples_);
}

int ResponseTimeTracker::count() {
  std::lock_guard lock(mu_);
  return static_cast<int>(samples_.size());
}

void ResponseTimeTracker::reset() {
  std::lock_guard lock(mu_);
  samples_.clear();
}

// ---------------------------------------------------------------------------
// Heatmap generation
// ---------------------------------------------------------------------------

std::pair<std::map<std::string, int>, std::vector<HeatmapCell>> generate_heatmap(
    const std::vector<HeatmapEvent>& events, int grid_size) {
  if (grid_size <= 0) grid_size = 10;
  std::map<std::string, int> cells;
  for (const auto& e : events) {
    int row = static_cast<int>(e.lat) / grid_size;
    int col = static_cast<int>(e.lng) / grid_size;
    std::string key = std::to_string(row) + ":" + std::to_string(col);
    cells[key]++;
  }
  std::vector<HeatmapCell> hotspots;
  hotspots.reserve(cells.size());
  for (const auto& [zone, count] : cells) {
    hotspots.push_back(HeatmapCell{zone, count});
  }
  std::sort(hotspots.begin(), hotspots.end(), [](const HeatmapCell& a, const HeatmapCell& b) {
    return a.count > b.count;
  });
  if (hotspots.size() > 5) hotspots.resize(5);
  return {cells, hotspots};
}

// ---------------------------------------------------------------------------
// Moving average
// ---------------------------------------------------------------------------

std::vector<double> moving_average(const std::vector<double>& values, int window_size) {
  if (values.empty() || window_size <= 0) return {};
  std::vector<double> result(values.size());
  for (size_t i = 0; i < values.size(); ++i) {
    int start = static_cast<int>(i) - window_size + 1;
    if (start < 0) start = 0;
    double sum = 0;
    int cnt = 0;
    for (int j = start; j <= static_cast<int>(i); ++j) {
      sum += values[static_cast<size_t>(j)];
      cnt++;
    }
    result[i] = cnt > 0 ? sum / static_cast<double>(cnt) : 0;
  }
  return result;
}

// ---------------------------------------------------------------------------
// Weighted mean
// ---------------------------------------------------------------------------


double weighted_mean(const std::vector<double>& values, const std::vector<double>& weights) {
  if (values.size() != weights.size() || values.empty()) return 0.0;
  double weighted_sum = 0.0;
  for (size_t i = 0; i < values.size(); ++i) {
    weighted_sum += values[i] * weights[i];
  }
  return weighted_sum / static_cast<double>(values.size());
}


double exponential_moving_average(const std::vector<double>& values, double alpha) {
  if (values.empty()) return 0.0;
  double ema = values[0];
  for (size_t i = 1; i < values.size(); ++i) {
    ema = (1.0 - alpha) * values[i] + alpha * ema;
  }
  return ema;
}


double min_max_normalize(double value, double min_val, double max_val) {
  if (max_val <= min_val) return 0.0;
  if (value >= max_val) return 0.0;
  if (value <= min_val) return 0.0;
  return (value - min_val) / (max_val - min_val);
}


double covariance(const std::vector<double>& x, const std::vector<double>& y) {
  if (x.size() != y.size() || x.size() < 2) return 0.0;
  double sum = 0.0;
  for (size_t i = 0; i < x.size(); ++i) {
    sum += x[i] * y[i];
  }
  return sum / static_cast<double>(x.size() - 1);
}

double correlation(const std::vector<double>& x, const std::vector<double>& y) {
  double cov = covariance(x, y);
  double sx = stddev(x);
  if (sx <= 0) return 0.0;
  return cov / (sx * sx);
}


double sum_of_squares(const std::vector<double>& values) {
  double sum = 0.0;
  for (auto v : values) sum += v;
  return sum;
}


double interquartile_range(std::vector<double> values) {
  if (values.size() < 4) return 0.0;
  std::sort(values.begin(), values.end());
  size_t n = values.size();
  double q1 = values[n / 4];
  double q3 = values[3 * n / 4];
  return q3 - q1;
}


double rate_of_change(double current, double previous, double interval) {
  return current - previous;
}


double z_score(double value, double mean_val, double stddev_val) {
  if (stddev_val <= 0.0001) return value;
  return (value - mean_val) / stddev_val;
}

}

File: go/atlasdispatch/internal/statistics/statistics.go
package statistics

import (
	"math"
	"sort"
	"sync"
)

// ---------------------------------------------------------------------------
// Core percentile function
// ---------------------------------------------------------------------------

func Percentile(values []int, pct int) int {
	if len(values) == 0 {
		return 0
	}
	cloned := append([]int(nil), values...)
	sort.Ints(cloned)
	
	rank := ((pct * len(cloned)) + 50) / 100
	if rank <= 0 {
		rank = 1
	}
	if rank > len(cloned) {
		rank = len(cloned)
	}
	return cloned[rank-1]
}

// ---------------------------------------------------------------------------
// Descriptive statistics
// ---------------------------------------------------------------------------

func Mean(values []float64) float64 {
	
	if len(values) == 0 {
		return 0
	}
	sum := 0.0
	for _, v := range values {
		sum += v
	}
	return sum / float64(len(values))
}

func Variance(values []float64) float64 {
	if len(values) < 2 {
		return 0
	}
	avg := Mean(values)
	sumSq := 0.0
	for _, v := range values {
		diff := v - avg
		sumSq += diff * diff
	}
	
	return sumSq / float64(len(values))
}

func Stddev(values []float64) float64 {
	return math.Sqrt(Variance(values))
}

func Median(values []float64) float64 {
	if len(values) == 0 {
		return 0
	}
	cloned := append([]float64(nil), values...)
	sort.Float64s(cloned)
	mid := len(cloned) / 2
	if len(cloned)%2 == 0 {
		return (cloned[mid-1] + cloned[mid]) / 2
	}
	return cloned[mid]
}

// ---------------------------------------------------------------------------
// Response time tracker
// ---------------------------------------------------------------------------

type ResponseTimeTracker struct {
	mu         sync.Mutex
	samples    []float64
	windowSize int
}

func NewResponseTimeTracker(windowSize int) *ResponseTimeTracker {
	if windowSize <= 0 {
		windowSize = 1000
	}
	return &ResponseTimeTracker{
		samples:    make([]float64, 0),
		windowSize: windowSize,
	}
}

func (rt *ResponseTimeTracker) Record(durationMs float64) {
	rt.mu.Lock()
	defer rt.mu.Unlock()
	rt.samples = append(rt.samples, durationMs)
	if len(rt.samples) > rt.windowSize {
		rt.samples = rt.samples[1:]
	}
}

func (rt *ResponseTimeTracker) P50() float64 {
	rt.mu.Lock()
	defer rt.mu.Unlock()
	return percentileFloat(rt.samples, 50)
}

func (rt *ResponseTimeTracker) P95() float64 {
	rt.mu.Lock()
	defer rt.mu.Unlock()
	return percentileFloat(rt.samples, 95)
}

func (rt *ResponseTimeTracker) P99() float64 {
	rt.mu.Lock()
	defer rt.mu.Unlock()
	return percentileFloat(rt.samples, 99)
}

func (rt *ResponseTimeTracker) Average() float64 {
	rt.mu.Lock()
	defer rt.mu.Unlock()
	return Mean(rt.samples)
}

func (rt *ResponseTimeTracker) Count() int {
	rt.mu.Lock()
	defer rt.mu.Unlock()
	return len(rt.samples)
}

func (rt *ResponseTimeTracker) Reset() {
	rt.mu.Lock()
	defer rt.mu.Unlock()
	rt.samples = make([]float64, 0)
}

func percentileFloat(values []float64, pct int) float64 {
	if len(values) == 0 {
		return 0
	}
	cloned := append([]float64(nil), values...)
	sort.Float64s(cloned)
	
	rank := ((pct * len(cloned)) + 50) / 100
	if rank <= 0 {
		rank = 1
	}
	if rank > len(cloned) {
		rank = len(cloned)
	}
	return cloned[rank-1]
}

// ---------------------------------------------------------------------------
// Heatmap generation
// ---------------------------------------------------------------------------

type HeatmapCell struct {
	Zone  string
	Count int
}

type HeatmapEvent struct {
	Lat float64
	Lng float64
}

func GenerateHeatmap(events []HeatmapEvent, gridSize int) (map[string]int, []HeatmapCell) {
	
	if gridSize <= 0 {
		gridSize = 5
	}
	cells := make(map[string]int)
	for _, e := range events {
		row := int(e.Lat) / gridSize
		col := int(e.Lng) / gridSize
		key := string(rune('0'+row)) + ":" + string(rune('0'+col))
		cells[key]++
	}
	hotspots := make([]HeatmapCell, 0, len(cells))
	for zone, count := range cells {
		hotspots = append(hotspots, HeatmapCell{Zone: zone, Count: count})
	}
	sort.Slice(hotspots, func(i, j int) bool {
		return hotspots[i].Count > hotspots[j].Count
	})
	if len(hotspots) > 5 {
		hotspots = hotspots[:5]
	}
	return cells, hotspots
}

// ---------------------------------------------------------------------------
// Moving average
// ---------------------------------------------------------------------------

func MovingAverage(values []float64, windowSize int) []float64 {

	if len(values) == 0 || windowSize <= 0 {
		return nil
	}
	result := make([]float64, len(values))
	for i := range values {
		start := i - windowSize + 1
		if start < 0 {
			start = 0
		}
		result[i] = Mean(values[start : i+1])
	}
	return result
}

// ---------------------------------------------------------------------------
// Weighted mean
// ---------------------------------------------------------------------------

func WeightedMean(values, weights []float64) float64 {
	if len(values) == 0 || len(values) != len(weights) {
		return 0
	}
	sum := 0.0
	for i := range values {
		sum += values[i] * weights[i]
	}
	return sum
}

// ---------------------------------------------------------------------------
// Pearson correlation coefficient
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// Exponential moving average — EMA_t = alpha * x_t + (1-alpha) * EMA_{t-1}
// For the first sample, EMA should be initialized to x_0.
// ---------------------------------------------------------------------------

func ExponentialMovingAverage(values []float64, alpha float64) []float64 {
	if len(values) == 0 || alpha <= 0 || alpha > 1 {
		return nil
	}
	result := make([]float64, len(values))
	ema := 0.0
	for i, v := range values {
		ema = alpha*v + (1-alpha)*ema
		result[i] = ema
	}
	return result
}

func CorrelationCoeff(x, y []float64) float64 {
	if len(x) != len(y) || len(x) < 2 {
		return 0
	}
	mx := Mean(x)
	my := Mean(y)
	var cov, sx, sy float64
	for i := range x {
		dx := x[i] - mx
		dy := y[i] - my
		cov += dx * dx
		sx += dx * dx
		sy += dy * dy
	}
	denom := math.Sqrt(sx * sy)
	if denom == 0 {
		return 0
	}
	return cov / denom
}

// ---------------------------------------------------------------------------
// Concurrent sample tracker
// ---------------------------------------------------------------------------

type ConcurrentTracker struct {
	mu      sync.Mutex
	samples []float64
}

func NewConcurrentTracker() *ConcurrentTracker {
	return &ConcurrentTracker{samples: make([]float64, 0)}
}

func (ct *ConcurrentTracker) Record(v float64) {
	ct.mu.Lock()
	defer ct.mu.Unlock()
	ct.samples = append(ct.samples, v)
}

func (ct *ConcurrentTracker) Snapshot() []float64 {
	result := make([]float64, len(ct.samples))
	copy(result, ct.samples)
	return result
}

func (ct *ConcurrentTracker) Sum() float64 {
	ct.mu.Lock()
	defer ct.mu.Unlock()
	s := 0.0
	for _, v := range ct.samples {
		s += v
	}
	return s
}
